File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/HandDealer.java
package com.panopset.blackjackEngine;

public class HandDealer extends Hand {

  public HandDealer(HandDealer dealer) {
    this();
    for (BlackjackCard card : dealer.cards) {
      this.cards.add(card);
    }
  }

  public HandDealer() {
    super();
  }

  public synchronized BlackjackCard getUpCard() {
    return cards.get(1);
  }

  public synchronized BlackjackCard getDownCard() {
    return cards.get(0);
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/HandDealer.kt
package com.panopset.blackjackEngine

class HandDealer() {

    private val hand = Hand()

    constructor(dealer: HandDealer) : this() {
        val iterator = dealer.hand.getIterator()
        val cardsToAddToHand = ArrayList<BlackjackCard>()
        while (iterator.hasNext()) {
            val card = iterator.next()
            cardsToAddToHand.add(card)
        }
        for (card in cardsToAddToHand) {
            hand.cards.add(card)
        }
    }

    fun getCards(): MutableList<BlackjackCard>  {
        return hand.cards
    }

    fun isFinal(): Boolean {
        return hand.isFinal
    }

    fun hasCards(): Boolean {
        return hand.hasCards()
    }

    fun dealCard(blackjackCard: BlackjackCard) {
        hand.dealCard(blackjackCard)
    }

    fun getHandValue(): Int {
        return hand.value
    }

    fun getFirstCard(): BlackjackCard {
        return hand.cards[0]
    }

    fun getSecondCard(): BlackjackCard {
        return hand.cards[1]
    }

    fun isSoft(): Boolean {
        return hand.isSoft
    }

    fun stand() {
        hand.stand()
    }

    fun isNatural21(): Boolean {
        return hand.isNatural21()
    }

    val upCard: BlackjackCard
        get() = hand.cards[1]

    val downCard: BlackjackCard
        get() = hand.cards[0]
}
